ci/ray_ci/automation/docker_tags_lib.py

```diff
@@ -169,6 +176,88 @@ def delete_tag(tag: str, docker_hub_token: str) -> bool:
     return True
 
 
+def _is_release_tag(
+    tag: str,
+    release_versions: Optional[List[str]] = None,
+) -> bool:
+    """
+    Check if tag is a release tag & is in the list of release versions.
+    Args:
+        tag: Docker tag name
+        release_versions: List of release versions.
+            If None, don't filter by release version.
+    Returns:
+        True if tag is a release tag and is in the list of release versions.
+            False otherwise.
+    """
+    variables = tag.split(".")
+    if len(variables) != 3 and "post1" not in tag:
+        return False
+    if not variables[0].isnumeric() or not variables[1].isnumeric():
+        return False
+    if (
+        not variables[2].isnumeric()
+        and "rc" not in variables[2]
+        and "-" not in variables[2]
+    ):
+        return False
+
+    if "-" in variables[2]:
+        variables[2] = variables[2].split("-")[0]
+    release_version = ".".join(variables)
+    if release_versions and release_version not in release_versions:
+        return False
+
+    return True
+
+
+def _call_crane_cp(tag: str, source: str, aws_ecr_repo: str):
+    try:
+        with subprocess.Popen(
+            [
+                "crane",
+                "cp",
+                source,
+                f"{aws_ecr_repo}:{tag}",
+            ],
+            stdout=subprocess.PIPE,
+            text=True,
+        ) as proc:
+            output = ""
+            for line in proc.stdout:
+                logger.info(line + "\n")
+                output += line
+            return_code = proc.wait()
+            if return_code:
+                raise subprocess.CalledProcessError(return_code, proc.args)
+            return output
+    except subprocess.CalledProcessError as e:
+        return f"Error: {e.output}"
```

Comment:
perhaps return a pair of (exit_code, msg) (rather then encoding error case as a string 'Error' in the msg)

Reply:
converting `crane cp` would be pretty challenging.

as discussed in other places:

- use the bazel `http_archive` to import the crane release archive file into the bazel workspace
- use bazel runfiles to add the `crane` binary inside the archive file into the data part of your python script
- when running `crane` via subprocess, but with the binary (first arg) pointing to the data part.

Reply:
I just updated implementation to use crane binary from bazel runfile

Reply:
[1]